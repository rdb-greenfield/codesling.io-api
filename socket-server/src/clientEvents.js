import axios from 'axios';

import { success } from './lib/log';
import {
  serverInitialState,
  clientOneServerChanged,
  clientTwoServerChanged,
  serverLeave,
  serverRun,
  serverMessage,
} from './serverEvents';
import _ from 'lodash';

/**
 *
 *  Client emissions (server listeners)
 *
 *  more on socket emissions:
 *  @url {https://socket.io/docs/emit-cheatsheet/}
 *
 *  @param room is an ES6 Map, containing { id, state }
 *  @url {https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map}
 *
 */
const clientReady = ({ io, client, room }, payload) => {
  // payload contains the challenge object
  success('client ready heard');
  serverInitialState({ io, client, room }, payload);
};

const clientOneUpdate = ({ io, client, room }, payload) => {
  const { text, player } = payload;
  success('client update heard. payload.text = ', payload);
  room.set('playerOne.text', text);
  clientOneServerChanged({
    io,
    client,
    room,
    player,
  });
};

const clientTwoUpdate = ({ io, client, room }, payload) => {
  const { text, player } = payload;
  success('client update heard. payload.text = ', payload);
  room.set('playerTwo.text', text);
  clientTwoServerChanged({
    io,
    client,
    room,
    player,
  });
};

const clientDisconnect = ({ io, room }) => {
  success('client disconnected');
  serverLeave({ io, room });
};

const clientRun = async ({ io, room }, payload) => {
  success('running code from client. room.get("text") = ', room.get('text'));
  const {
    text, player, tests, fnName,
  } = payload;
  const url = process.env.CODERUNNER_SERVICE_URL;
  const cases = tests.split('\n');
  const tuples = [];
  for (let i = 0; i < cases.length; i += 2) {
    tuples.push([cases[i], cases[i + 1]]);
  }
  try {
    const { data } = await axios.post(`${url}/submit-code`, { code: text });
    let stdout = data;
    if (fnName) {
      let allPass = true;
      for (let i = 0; i < tuples.length; i++) {
        const testCase = `${text}\n${fnName}(${tuples[i][0]});`;
        const result = await axios.post(`${url}/submit-code`, { code: testCase });
        if (allPass) {
          allPass = _.isEqual(JSON.parse(result.data.result), JSON.parse(tuples[i][1]));
        }
      }
      if (allPass) {
        stdout = { result: 'WINNER!' };
      }
    }
    serverRun({ io, room }, { stdout, player });
  } catch (e) {
    success('error posting to coderunner service from socket server. e = ', e);
  }
};

const clientMessage = async ({ io, room }, payload) => {
  success('client message heard');
  const url = process.env.REST_SERVER_URL;
  try {
    const { data } = await axios.post(`${url}/messages/`, payload);
    serverMessage({ io, room }, data);
  } catch (e) {
    success('error saving message to the database. e = ', e);
  }
};

const clientEmitters = {
  'client.ready': clientReady,
  'clientOne.update': clientOneUpdate,
  'clientTwo.update': clientTwoUpdate,
  'client.disconnect': clientDisconnect,
  'client.run': clientRun,
  'client.message': clientMessage,
};

export default clientEmitters;
