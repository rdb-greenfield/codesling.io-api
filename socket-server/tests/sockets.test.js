import ioClient from 'socket.io-client/dist/socket.io.js';
import delay from 'delay';
import randomId from 'random-id';
import dotenv from 'dotenv';
import { after } from 'lodash';

import clientEvents from '../src/clientEvents';

const text =
`function hello() {
  console.log('hello!');
}

hello();
`;

dotenv.config();
dotenv.load();

describe('Client interactions', () => {
  let client1;
  let client2;
  let roomId = randomId(10);

  beforeAll((done) => {
    done = after(2, done);
    const query = { query: `roomId=${roomId}` };
    client1 = ioClient.connect(`${process.env.HOST}:${process.env.PORT}`, query);
    client2 = ioClient.connect(`${process.env.HOST}:${process.env.PORT}`, query);
    client1.on('connect', () => {
      done();
    });
    client2.on('connect', () => {
      done();
    });
  });

  afterAll(async (done) => {
    client1.disconnect();
    client2.disconnect();
    await delay(300);
    done();
  });

  // Expects people to be able to connect to the socket-server
  test('Should add a new user to a room and receive the same text', async (done) => {
    done = after(2, done);
    expect.assertions(6);
    const handler = (payload) => {
      try {
        expect(payload.playerOneText).toMatchSnapshot();
        expect(payload.playerTwoText).toMatchSnapshot();
        // make sure real test above does not throw
        expect(true).toBe(true);
      } catch (e) {
        console.log(e.toString());
      }
      done();
    };

    client1.on('server.initialState', handler);
    client2.on('server.initialState', handler);

    client1.emit('client.ready', { challenge: '' });
    client2.emit('client.ready', { challenge: '' });
  });

  // Expects clients to be able to connect with one another
  test('Should be able to hear emissions from other clients', (done) => {
    expect.assertions(2);
    client2.on('serverOne.changed', (payload) => {
      try {
        expect(payload.text).toMatchSnapshot();
        expect(true).toBe(true);
      } catch (e) {
        console.log(e.toString());
      }
      done();
    });
    client1.emit('clientOne.update', { text: 'console.log("Hello, World!")' });
  });

  test('Should be able to run code and have both clients receive stdout', (done) => {
    done = after(2, done);
    expect.assertions();
    const serverRunHandler = ({ stdout }) => {
      try {
        expect(stdout).toMatchSnapshot();
        expect(true).toBe(true);
      } catch(e) {
        console.log(e.toString());
      }
      done();
    };
    client1.on('server.run', serverRunHandler);
    client2.on('server.run', serverRunHandler);
    client1.emit('client.run', { text, player: 1 });
  });

  // Expects clients to disconnect
  test('Client should have disconnected', (done) => {
    expect.assertions(1);
    client1.on('server.leave', () => {
      // ensures server.leave is heard for other room clients
      expect(true).toBe(true);
      done();
    });
    client2.emit('client.disconnect');
  });
});
