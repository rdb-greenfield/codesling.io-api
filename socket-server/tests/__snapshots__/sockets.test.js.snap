// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Client interactions Should add a new user to a room and receive the same text 1`] = `
"function hello() {
  console.log('hello!');
}

hello();
"
`;

exports[`Client interactions Should add a new user to a room and receive the same text 2`] = `
"function hello() {
  console.log('hello!');
}

hello();
"
`;

exports[`Client interactions Should add a new user to a room and receive the same text 3`] = `
"function hello() {
  console.log('hello!');
}

hello();
"
`;

exports[`Client interactions Should add a new user to a room and receive the same text 4`] = `
"function hello() {
  console.log('hello!');
}

hello();
"
`;

exports[`Client interactions Should be able to hear emissions from other clients 1`] = `"console.log(\\"Hello, World!\\")"`;

exports[`Client interactions Should be able to run code and have both clients receive stdout 1`] = `
Object {
  "console": Array [
    "hello!",
  ],
  "result": "null",
}
`;

exports[`Client interactions Should be able to run code and have both clients receive stdout 2`] = `
Object {
  "console": Array [
    "hello!",
  ],
  "result": "null",
}
`;
